app/src/main/java/de/westnordost/streetcomplete/util/SpatialCache.kt

```diff
@@ -0,0 +1,169 @@
+package de.westnordost.streetcomplete.util
+
+import android.util.Log
+import de.westnordost.streetcomplete.data.download.tiles.TilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilePos
+import de.westnordost.streetcomplete.data.download.tiles.enclosingTilesRect
+import de.westnordost.streetcomplete.data.download.tiles.minTileRect
+import de.westnordost.streetcomplete.data.osm.mapdata.BoundingBox
+import de.westnordost.streetcomplete.data.osm.mapdata.LatLon
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isCompletelyInside
+
+/**
+ * Spatial cache containing items of type T that each must have an id of type K and a position
+ * (LatLon). See [getKey] and [getPosition].
+ *
+ * @fetch needs get data from db and fill other caches, e.g. questKey -> Quest
+ *
+ * the bbox queried in get() must fit in cache, i.e. may not be larger than maxTiles at tileZoom.
+ */
+class SpatialCache<K, T>(
+    private val tileZoom: Int,
+    private val maxTiles: Int,
+    initialCapacity: Int?,
+    private val fetch: (BoundingBox) -> Collection<T>,
+    private val getKey: T.() -> K,
+    private val getPosition: T.() -> LatLon,
+) {
+    private val byTile = LinkedHashMap<TilePos, HashSet<T>>((maxTiles/0.75).toInt(), 0.75f, true)
+    private val byKey = initialCapacity?.let { HashMap<K, T>(it) } ?: HashMap<K, T>()
+    val keys: Set<K> get() = synchronized(this) { byKey.keys.toHashSet() } // return a new set!
```

Comment:
> However, iterating through that set is not thread-safe

Yes, this is why I wanted a new set.

The `HashSet` was because I thought it would be faster to create than the default set, which is a `LinkedHashSet`. I just did some short test, and `toHashSet` is indeed somewhat faster (~25%) than `toSet`.
`toList` is 10 times faster, so we could also do that...

Reply:
Interesting. Well I guess if the keys are only iterated and nothing else, then it can also be a list.

Actually, since this is not a simple field but each time it is called, a new list is created, I think it is more explicit if this were a function, not a field.

Reply:
Done, it's now a list returned by a documented function.